```{r include=FALSE}

knitr::opts_chunk$set(echo = TRUE, cache = TRUE)

# Load data
source("./scripts/0_meta_env_prep.R")

```

# Quality control, data exploration

## Quality control

Quality control plots revealed no problems with dataset quality.

### Library size

![Library size ranged between 30 million and 70 million and is sufficient for RNA-seq.](output/plots_QC/Library%20size.png){width="80%"}

### Counts distribution

![](output/plots_QC/distribution.png){width="100%"}

Library composition was highly similar between samples, both before and after TMM normalisation.

### Counts density, before and after removal of lowly-expressed genes

![](output/plots_QC/Density%20of%20count%20values%20-%20cDNA.png){width="100%"}

![](output/plots_QC/voom%20mean-variance%20trend.png){width="100%"}

Genes with low expression (determined using `edgeR::filterByExpr(min.count = 40)`) were successfully removed. After filtering, the voom mean-variance plot showed a downwards trend characteristic of filtered RNA-seq data.

## Data exploration

### PCA

```{r out.width = "100%"}

quant_DGE_clean <-
  readRDS(file = file.path("output",
                           "data_expression",
                           "pre_DGE",
                           "quant_cDNA_DGE_filter.RDS"))

glimmaMDS(quant_DGE_clean,
          width = 800,
          height = 500)

```

Suggested settings:

-   `x_axis`: dim1

-   `y_axis`: dim2

-   `colour_by`: treatment_combined

-   `shape_by`: genotype

PCA plot demonstrates:

-   No outliers except in `ATP7B-KO_Cu_trientine` group.
    -   This is likely caused by high rRNA content in one of the samples.
-   Strong clustering of samples by both treatment and genotype
    -   Differences between treatments are captured by PC1 (dim1 / x-axis)
        -   Untreated and Cu treated samples are on opposite sides of dim 1.
        -   Cu + trientine treated samples cluster very closely with untreated samples, suggesting highly similar gene expression profile
        -   The Cu + D-penicilamine samples cluster is closer to the untreated cluster than the Cu cluster.
    -   Differences between genotypes are captured by PC2 (dim2 / y-axis)
        -   The Cu treated group is further to the right in KO cells compared to WT cells. This suggest that differential expression in response to Cu is stronger in KO cells compared to WT cells.

### Heatmap

#### No clustering

![](output/plots_heatmap/normalised_expression_genes_unclustered.png){width="100%"}

#### With sample clustering

![](output/plots_heatmap/normalised_expression_genes_clustered.png){width="100%"}
