---
title: "HepG2 Transcriptome Report"
author: "Mitchell Nguyen"
date: "`r Sys.Date()`"
output: bookdown::gitbook
documentclass: book
geometry: margin=2cm
biblio-style: apalike
link-citations: yes
bibliography: references.bib
---

# Preface

## Methods

### Exploratory analysis and QC

Gene counts produced by the `nf-core/rnaseq` Nextflow pipeline were imported to R using tximeta and annotated using the org.Hs.eg.db Bioconductor package. Lowly expressed genes were filtered using the filterByExpr() function implemented in edgeR, with the minimum count threshold set to 40 counts. This filtering benefits the estimation of mean-variance relationship in the data, removes genes that were too lowly expressed to be accurately quantified, and reduces multiple testing burden [@10.12688/f1000research.9005.3]. The trimmed mean of M-values (TMM) method was used to normalise all gene counts for library size and count biases between samples.

*NB: I wasn't able to extract all parameters from the Nextflow logs. Based on the result files, I believe that the RNA-seq reads were aligned to GRCh37 using STAR and quantified using RSEM.*

### Data exploration

Unsupervised clustering using multidimensional scaling (MDS) plots implemented in `Glimma` was used for data exploration. In this context, MDS reduces the dimensionality of our datasets from thousands of genes to a smaller number of dimensions. Here, each dimension is uncorrelated from each other, the first dimensions (dim1, dim2, etc.) are composed of the most highly variable genes in the dataset, and together, the dimensions retain most of the original datasetâ€™s variance. Additionally, expression heatmaps was used to summarise differential expression in the dataset. Heatmaps visualised only differentially expressed genes for clarity.

### Differential gene expression (DGE) analysis

The normalised transcriptome data was used to build a model in limma-voom [@ritchie2015] with the necessary experimental conditions (genotype, treatment) coded in the model matrix. The voomLmFit function in edgeR was used to convert RNA-seq counts to logCPM, estimate the mean-variance relationship in the dataset, and linear model fit . A contrast matrix was used to define desired comparisons between genotypes and treatments and timepoints to discover differentially expressed genes (DEGs) for each comparison. Comparisons were made to determine (a) the effect of ATP7B knockout status on gene expression (b) the effect of treatment of cells with copper and copper chelators on gene expression, and (c) the effect of of ATP7B knockout status on gene expression changes in response to copper and chelators treatment. Genes with FDR \< 0.05 were considered as differentially expressed genes (DEGs).

#### Model matrix specifications

```{r echo=FALSE}

fit_contrasts <- readRDS(file.path("output",
                                   "data_expression",
                                   "post_DGE",
                                   "fit_contrasts.RDS"))

fit_contrasts[["design"]]
```

#### Contrast matrix specifications

```{r echo=FALSE}

fit_contrasts[["contrasts"]]
```

#### List of comparisons

| Comparison           | logFC formula                                                                |
|------------------|------------------------------------------------------|
| KO_vs_WT_UT          | $logFC_{KO \_ vs \_ WT \_ UT} = {Untreated}_{ATP7B-KO} - {Untreated}_{WT}$   |
| CU_vs_UT_WT          | $logFC_{CU \_ vs \_ UT \_ WT} = {Cu}_{WT} - {Untreated}_{WT}$                |
| DPEN_vs_CU_WT        | $logFC_{DPEN \_ vs \_ CU \_ WT} = {Cu \_ DPEN}_{WT} - {Cu}_{WT}$             |
| TRIEN_vs_Cu_WT       | $logFC_{TRIEN \_ vs \_ CU \_ WT} = {Cu \_ TRIEN}_{WT} - {Cu}_{WT}$           |
| DPEN_vs_UT_WT        | $logFC_{DPEN \_ vs \_ UT \_ WT} = {Cu \_ DPEN}_{WT} - {Untreated}_{WT}$      |
| TRIEN_vs_UT_WT       | $logFC_{TRIEN \_ vs \_ UT \_ WT} = {Cu \_ TRIEN}_{WT} - {Untreated}_{WT}$    |
| CU_vs_UT_KO          | $logFC_{CU \_ vs \_ UT \_ KO} = {Cu}_{KO} - {Untreated}_{KO}$                |
| DPEN_vs_CU_KO        | $logFC_{DPEN \_ vs \_ CU \_ KO} = {Cu \_ DPEN}_{KO} - {Cu}_{KO}$             |
| TRIEN_vs_CU_KO       | $logFC_{TRIEN \_ vs \_ CU \_ KO} = {Cu \_ TRIEN}_{KO} - {Cu}_{KO}$           |
| DPEN_vs_UT_KO        | $logFC_{DPEN \_ vs \_ UT \_ KO} = {Cu \_ DPEN}_{KO} - {Untreated}_{KO}$      |
| TRIEN_vs_UT_KO       | $logFC_{TRIEN \_ vs \_ UT \_ KO} = {Cu \_ TRIEN}_{KO} - {Untreated}_{KO}$    |
| CU_vs_UT_KO_vs_WT    | $logFC = logFC_{CU \_ vs \_ UT \_ KO} - logFC_{CU \_ vs \_ UT \_ WT}$        |
| DPEN_vs_CU_KO_vs_WT  | $logFC = logFC_{DPEN \_ vs \_ CU \_ KO} - logFC_{DPEN \_ vs \_ CU \_ WT}$    |
| TRIEN_vs_CU_KO_vs_WT | $logFC = logFC_{TRIEN \_ vs \_ CU \_ KO}  - logFC_{TRIEN \_ vs \_ CU \_ WT}$ |
| DPEN_vs_UT_KO_vs_WT  | $logFC = logFC_{DPEN \_ vs \_ UT \_ KO} - logFC_{DPEN \_ vs \_ UT \_ WT}$    |
| TRIEN_vs_UT_KO_vs_WT | $logFC = logFC_{TRIEN \_ vs \_ UT \_ KO} - logFC_{TRIEN \_ vs \_ UT \_ WT}$  |

### Visualisation of DEGs

#### Volcano plots

Volcano plots implemented in the Glimma package were used for visualising changes across the whole genome between 2 conditions, with the x-axis representing log-fold change between selected conditions and the y-axis representing $-log_{10}(p.value)$.

#### 3D volcano plots

3D cylindrical volcano plots implemented in the Volcano3D package (Lewis et al., 2019) were used for visualising changes across the whole genome between 3 conditions. In this plot, gene expression (as logCPM) was projected on to a two-dimensional polar coordinate system and coloured based on the condition where they were considered as significantly upregulated. The vertical axis represents $-log_{10}(p.value)$.

#### XY plot

XY plots were used for visualising correlation in gene expression changes in response to treatments between WT and ATP7B-KO cells. In this this plot, log-fold change for WT cells were plotted on the x-axis and log-fold change for KO cells were plotted on the y-axis. Colors were used to colour genes based on statistical significance of differential expression (Blue: significant in WT only; Red: significant in KO only; Purple: significant in both).

### Functional enrichment analysis

Functional enrichment analysis describes a class of bioinformatics methods that identifies biological functions significantly overrepresented in a list of genes. Multiple gene set databases are employed in this study to gauge a complete picture of biological changes across the experimental conditions. Information from these databases were used to determine functional enrichment by the camera algorithm [@wu2012].

#### Gene set databases

Table (x) lists the gene set databases used in our study. Two main databases utilised were MSigDB Hallmark and Gene Ontology. All gene sets were downloaded from the Broad Institute Molecular Signatures Database version 2023.2 as GMT files and imported to the R environment.

| Source        | DL from | Name                                  |
|---------------|---------|---------------------------------------|
| MSigDB        | MSigDB  | H:                                    |
| Gene Ontology | MSigDB  | C5/GO/BP: Biological Process ontology |
| Gene Ontology | MSigDB  | C5/GO/CC: Cellular Component ontology |
| Gene Ontology | MSigDB  | C5/GO/MF: Molecular Function ontology |

#### Correlation adjusted mean rank (camera) gene set enrichment analysis

A gene set enrichment-type analysis was performed using camera with the gene sets described above. This method was used to perform functional enrichment on gene lists derived directly from differential expression analysis, with the primary benefit of being able to detect functional enrichment when differential expression was not significant. Camera improves on previously implemented gene set enrichment analysis algorithms by correcting for inter-gene correlation to prevent false positives. The test was performed with the inter-gene correlation factor set to 0.01 (as suggested by the original authors); gene sets with FDR \< 0.05 were considered significantly enriched.
