# Preface

## Methods

*TODO: Replace model matrix and contrast matrix with relevant DESeq2 specs*

### RNA extraction and targeted transcriptome sequencing

Targeted transcriptome sequencing was performed using established lab protocols. In brief, total RNA was harvested at key time points (Passage 5, 7, 13, at culture days 3 and 5, ± Heparin 10µg/mL) using the Direct-zol RNA MiniPrep kit (Zymo Research, USA, Cat no. R2050). Sequencing library preparation was performed using the Ion AmpliSeq™ Transcriptome Human Gene Expression Kit (Ion Torrent, USA, Cat no. A26325) and Ion Xpress Barcode Adapters (Ion Torrent, USA, Cat no. 4471250) following the manufacturer’s protocol. This AmpliSeq kit utilises a multiplex PCR primer pool targeting over 20,000 human RefSeq genes (18,574 mRNAs, 2,228 ncRNAs), generating a single amplicon from each gene for sequencing and quantitation (Papp et al., 2018). Quality control was performed using the 2100 Bioanalyser (Agilent, USA, Cat no. G2939BA) on the final transcriptome library. Each barcoded amplicon was then diluted, mixed into an equimolar library pool, then loaded into an Ion Chef System (Ion Torrent, USA) and sequenced using an Ion Proton semi-conductor sequencer (Ion Torrent, USA).

### Bioinformatics

#### Quality control, pseudoalignment, filtering

Raw binary alignment metric (BAM) sequence files were downloaded from the Ion Torrent server, converted to FASTQ files using SAMtools, then compressed with bgzip for storage. Quality control was performed with FastQC to determine sequence quality. BBMap (Bushnell, 2014) was used to retain sequences between 50 bp and 150 bp containing reads of single amplicons generated by the multiplex PCR reaction. Salmon (Patro et al., 2017) was used to align and quantify sequencing reads. In detail, the entire human mRNA sequence (release 109) was downloaded from the ENSEMBL database (https://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/). This sequence file was used to build a Salmon quasi-index that includes transcripts for protein-coding genes, with the entire human genome DNA sequence used as decoy, following the salmon package documentation. Sequenced reads were quasi-mapped to this reference index then quantified. Result files were imported to R using tximeta, annotated using the AnnotationDbi Bioconductor package for ENSEMBL version 109, then summarised from the transcript level to the gene level. Lowly expressed genes were filtered using the filterByExpr() function implemented in edgeR, with the minimum count threshold set to 40 counts. This filtering benefits the estimation of mean-variance relationship in the data, removes genes that were too lowly expressed to be accurately quantified, and reduces multiple testing burden (Law et al., 2018). The trimmed mean of M-values (TMM) method to adjust all gene counts for library size and count biases between samples.

#### Data exploration

Unsupervised clustering using PCA plots implemented in `PCAtools` was used for data exploration. In this context, MDS reduces the dimensionality of our datasets from thousands of genes to a smaller number of dimensions. Here, each dimension is uncorrelated from each other, the first dimensions (dim1, dim2, etc.) are composed of the most highly variable genes in the dataset, and together, the dimensions retain most of the original dataset’s variance. Additionally, expression heatmaps was used to summarise differential expression in the dataset. Heatmaps visualised only differentially expressed genes for clarity.

### Differential gene expression (DGE) analysis

The normalised transcriptome data was used to build a model in limma-voom [@ritchie2015] with the necessary experimental conditions (genotype, treatment) coded in the model matrix. The voomLmFit function in edgeR was used to convert RNA-seq counts to logCPM, estimate the mean-variance relationship in the dataset, and linear model fit . A contrast matrix was used to define desired comparisons between treatments and timepoints to discover differentially expressed genes (DEGs) for each comparison. Comparisons were made to [placeholder]. Genes with FDR \< 0.05 were considered as differentially expressed genes (DEGs).

#### Model matrix specifications

```{r} 
#| echo: False
fit_contrasts <- readRDS(file.path(
  "output",
  "data_expression",
  "post_DGE",
  "fit_contrasts.RDS"
))

fit_contrasts[["design"]]
```

#### Contrast matrix specifications

```r 
#| echo: False
fit_contrasts[["contrasts"]]
```

#### List of comparisons

### Visualisation of DEGs

#### Volcano plots

Volcano plots implemented in the Glimma package were used for visualising changes across the whole genome between 2 conditions, with the x-axis representing log-fold change between selected conditions and the y-axis representing $-log_{10}(p.value)$.

#### 3D volcano plots

3D cylindrical volcano plots implemented in the Volcano3D package (Lewis et al., 2019) were used for visualising changes across the whole genome between 3 conditions. In this plot, gene expression (as logCPM) was projected on to a two-dimensional polar coordinate system and coloured based on the condition where they were considered as significantly upregulated. The vertical axis represents $-log_{10}(p.value)$.

### Functional enrichment analysis

Functional enrichment analysis describes a class of bioinformatics methods that identifies biological functions significantly overrepresented in a list of genes. Multiple gene set databases are employed in this study to gauge a complete picture of biological changes across the experimental conditions. Information from these databases were used to determine functional enrichment by the camera algorithm [@wu2012].

#### Gene set databases

Table (x) lists the gene set databases used in our study. Two main databases utilised were MSigDB Hallmark and Gene Ontology. All gene sets were downloaded from the Broad Institute Molecular Signatures Database version 2023.2 as GMT files and imported to the R environment.

| Source        | DL from | Name                                  |
|---------------|---------|---------------------------------------|
| MSigDB        | MSigDB  | H:                                    |
| Gene Ontology | MSigDB  | C5/GO/BP: Biological Process ontology |
| Gene Ontology | MSigDB  | C5/GO/CC: Cellular Component ontology |
| Gene Ontology | MSigDB  | C5/GO/MF: Molecular Function ontology |

#### Correlation adjusted mean rank (camera) gene set enrichment analysis

A gene set enrichment-type analysis was performed using camera with the gene sets described above. This method was used to perform functional enrichment on gene lists derived directly from differential expression analysis, with the primary benefit of being able to detect functional enrichment when differential expression was not significant. Camera improves on previously implemented gene set enrichment analysis algorithms by correcting for inter-gene correlation to prevent false positives. The test was performed with the inter-gene correlation factor set to 0.01 (as suggested by the original authors); gene sets with FDR \< 0.05 were considered significantly enriched.
